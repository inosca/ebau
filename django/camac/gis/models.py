from django.db import models
from django.utils.translation import gettext_lazy as _


class GISConfig(models.Model):
    CLIENT_SOGIS = "camac.gis.clients.sogis.SoGisClient"
    CLIENT_PARAM = "camac.gis.clients.param.ParamClient"

    CLIENT_CHOICES = [
        (CLIENT_SOGIS, _("GIS Canton Solothurn")),
        (CLIENT_PARAM, _("Parameter")),
    ]

    slug = models.SlugField(primary_key=True)
    client = models.CharField(max_length=255, choices=CLIENT_CHOICES)
    config = models.JSONField(default=dict)
    disabled = models.BooleanField(default=False)
