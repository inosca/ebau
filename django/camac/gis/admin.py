from adminsortable2.admin import SortableAdminMixin
from django.contrib.admin import ModelAdmin, register
from django.db.models import JSONField
from django_json_widget.widgets import JSONEditorWidget

from camac.admin import EbauAdminMixin
from camac.gis.models import GISDataSource


@register(GISDataSource)
class GISDataSourceAdmin(EbauAdminMixin, SortableAdminMixin, ModelAdmin):
    list_display = ["description", "client", "disabled"]
    list_per_page = 50
    search_fields = ["description", "client"]
    formfield_overrides = {
        JSONField: {"widget": JSONEditorWidget},
    }
