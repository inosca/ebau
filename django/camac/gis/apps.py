from django.apps import AppConfig


class GisConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "camac.gis"
