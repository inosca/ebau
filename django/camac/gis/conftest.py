import pytest
from syrupy.filters import props


@pytest.fixture
def mock_municipalities(mocker):
    def mock(names=[]):
        mocker.patch(
            "camac.caluma.extensions.data_sources.Municipalities.get_data",
            return_value=[[i + 1, {"de": name}] for i, name in enumerate(names)],
        )

    return mock


@pytest.fixture
def gis_snapshot(snapshot):
    return snapshot(exclude=props("cache"))
