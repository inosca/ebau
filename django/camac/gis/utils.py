from enum import Enum
from typing import Any, List, Union

from django.utils.translation import gettext as _


class MergeStrategy(Enum):
    MERGE_FIRST = "merge_first"

    # actually not merge anything at all, give full control over return value
    # to each client (later clients can change output of previous clients)
    OVERRIDE = "override"


def get_bbox(x: Union[float, str], y: Union[float, str], buffer: int = 0) -> str:
    delta = 0

    if buffer:
        delta = buffer / 2

    try:
        x = float(x)
        y = float(y)
    except ValueError:
        raise ValueError(_("Coordinates must be floats"))

    return ",".join(map(str, [x - delta, y - delta, x + delta, y + delta]))


def to_query(query_params: dict) -> str:
    return "&".join(
        [f"{key}={value}" for key, value in query_params.items() if value is not None]
    )


def cast(value: Any, type: str) -> Any:
    if value is None:
        return None

    try:
        if type == "float":
            return float(value)
        elif type == "integer":
            return int(value)
        elif type == "string":
            return str(value)
    except ValueError:
        return None

    return value


def concat_values(*parts: Any) -> str:
    """Clean, deduplicate and concatenate values.

    This function cleans out empty values, deduplicates, and concatenates them
    using a comma as separator. If there is no meaningful value, it returns none
    and if there is only one value it returns it without converting it to a
    string.
    """

    # Filter out empty values
    non_empty_parts = [part for part in parts if part not in [None, ""]]

    # Return None if there are only empty values
    if len(non_empty_parts) == 0:
        return None
    # Return single value without joining to preserve the data type
    elif len(non_empty_parts) == 1:
        value = non_empty_parts[0]

        if isinstance(value, str):
            value = value.strip()

        return value

    clean_parts = [str(part).strip() for part in non_empty_parts]
    deduped_parts = sorted(set(clean_parts), key=clean_parts.index)

    return ", ".join(deduped_parts).strip()


def merge_table(
    existing_rows: List[dict],
    new_row: dict,
    merge_strategy: MergeStrategy,
) -> List[dict]:
    if merge_strategy == MergeStrategy.MERGE_FIRST:
        merge_data(existing_rows[0], new_row, merge_strategy)

    return existing_rows


def merge_data(data: dict, new_data: dict, merge_strategy: MergeStrategy) -> dict:
    for key, value in new_data.items():
        if key in data:
            if isinstance(data[key], list):
                for row in value:
                    value = merge_table(data[key], row, merge_strategy)
            elif merge_strategy == MergeStrategy.MERGE_FIRST:
                # If a previous data source already returned a value for a
                # certain question we concat the new and old value
                value = concat_values(data[key], value)

        data[key] = value
