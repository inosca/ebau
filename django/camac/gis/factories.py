from factory import Faker, fuzzy
from factory.django import DjangoModelFactory

from camac.gis import models


class GISDataSourceFactory(DjangoModelFactory):
    description = Faker("text")
    client = fuzzy.FuzzyChoice(dict(models.GISDataSource.CLIENT_CHOICES).keys())
    config = {}
    disabled = False

    class Meta:
        model = models.GISDataSource
