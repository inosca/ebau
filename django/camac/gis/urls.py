from django.urls import re_path

from camac.gis import views

urlpatterns = [
    re_path(
        r"^data/(?:(?P<task_id>[\w-]+)/)?$",
        views.GISDataView.as_view(),
        name="gis-data",
    ),
    re_path(
        r"^apply$",
        views.GISApplyView.as_view(),
        name="gis-apply",
    ),
]
