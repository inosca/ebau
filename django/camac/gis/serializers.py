from dataclasses import dataclass
from typing import List

from caluma.caluma_form.api import save_answer
from caluma.caluma_form.models import Document, Question
from django.core.cache import cache
from django.utils.translation import gettext as _
from rest_framework import serializers
from rest_framework.exceptions import PermissionDenied, ValidationError

from camac.caluma.extensions.permissions import CustomPermission
from camac.instance.models import Instance


@dataclass
class GISApplyResponse:
    pk: None
    instance: Instance
    questions: List[str]


class GISApplySerializer(serializers.Serializer):
    instance = serializers.PrimaryKeyRelatedField(queryset=Instance.objects)
    cache = serializers.UUIDField(write_only=True)
    questions = serializers.ListField(child=serializers.CharField(), read_only=True)

    def validate_cache(self, value):
        if not cache.has_key(value):
            raise ValidationError(
                _("Cache for key %(key)s does not exist") % {"key": value}
            )

        return value

    def _write_row(self, form_slug, row_data):
        row_document = Document.objects.create(form_id=form_slug)

        for question_slug, value in row_data.items():
            self._write_answer(row_document, question_slug, value)

        return str(row_document.pk)

    def _write_answer(self, document, question_slug, value):
        try:
            question = Question.objects.get(pk=question_slug)
        except Question.DoesNotExist:  # pragma: no cover
            return False

        answer_value = value["value"]

        if question.type == Question.TYPE_TABLE:
            answer_value = [
                self._write_row(value["form"], row) for row in value["value"]
            ]
        elif question.type == Question.TYPE_MULTIPLE_CHOICE:
            answer_value = [opt["value"] for opt in answer_value]

        save_answer(
            question=question,
            document=document,
            user=self.context["request"].caluma_info.context.user,
            value=int(answer_value)
            if question.type == "integer"
            else float(answer_value)
            if question.type == "float"
            else answer_value,
            meta={"gis-value": answer_value},
        )

        return True

    def create(self, validated_data):
        data = cache.get(validated_data["cache"])
        document = validated_data["instance"].case.document
        written_questions = set()

        if not CustomPermission(self.context["request"]).has_camac_edit_permission(
            document, self.context["request"].caluma_info
        ):
            raise PermissionDenied()

        for question_slug, value in data.items():
            success = self._write_answer(document, question_slug, value)

            if success:
                written_questions.add(question_slug)

        return GISApplyResponse(
            pk=None,
            instance=validated_data["instance"],
            questions=written_questions,
        )
