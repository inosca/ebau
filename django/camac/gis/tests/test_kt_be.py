import pytest
from caluma.caluma_form.models import Question
from django.core.management import call_command
from django.urls import reverse
from rest_framework import status

from camac.gis.models import GISDataSource


@pytest.fixture
def be_data_sources(
    question_factory, question_option_factory, option_factory, settings
):
    call_command("loaddata", settings.ROOT_DIR("kt_bern/config/gis.json"))
    gis_questions = [
        ("nutzungszone", Question.TYPE_TEXT),
        ("ueberbauungsordnung", Question.TYPE_TEXT),
        ("gebiet-mit-archaeologischen-objekten", Question.TYPE_CHOICE, ["ja", "nein"]),
        ("belasteter-standort", Question.TYPE_CHOICE, ["ja", "nein"]),
        ("handelt-es-sich-um-ein-baudenkmal", Question.TYPE_CHOICE, ["ja", "nein"]),
        ("gebiet-mit-naturgefahren", Question.TYPE_CHOICE, ["ja", "nein"]),
        (
            "objekt-des-besonderen-landschaftsschutzes",
            Question.TYPE_CHOICE,
            ["ja", "nein"],
        ),
        ("naturschutz", Question.TYPE_CHOICE, ["ja", "nein"]),
        (
            "gewaesserschutzbereich-v2",
            Question.TYPE_MULTIPLE_CHOICE,
            ["ueb", "ao", "au"],
        ),
        (
            "grundwasserschutzzonen-v2",
            Question.TYPE_MULTIPLE_CHOICE,
            [
                "s1",
                "s2",
                "s3zu",
                "sh",
                "sm",
                "sa",
                "sbw",
            ],
        ),
    ]
    for config in gis_questions:
        slug = config[0]
        type = config[1]
        q = question_factory(slug=slug, type=type, label=slug)
        if len(config) == 3:
            for option in config[2]:
                question_option_factory(
                    question=q,
                    option=option_factory(slug=f"{slug}-{option}", label=option),
                )

    return GISDataSource.objects.all()


# TODO: Figure out queued and concurrent testing, django_q sync=True is still broken.
@pytest.mark.parametrize(
    "egrids",
    [
        "CH673533354667",
        "CH643546955207",
        "CH851446093521",
        "CH673533354667,CH643546955207",
        "CH673533354667,CH851446093521",
        "CH643546955207,CH851446093521",
        "CH643546955207,CH843546955632",
        "CH643546955207,CH851446093521,CH673533354667",
    ],
)
@pytest.mark.vcr()
def test_be_client(
    db,
    admin_client,
    snapshot,
    vcr_config,
    egrids,
    be_data_sources,
):
    response = admin_client.get(
        reverse("gis-data"),
        data={
            "egrids": egrids,
        },
    )

    assert response.status_code == status.HTTP_200_OK
    snapshot.assert_match(response.json())


@pytest.mark.parametrize(
    "egrids",
    [
        "doesntexist",
        "emptypolygon",
        "emptygis",
    ],
)
@pytest.mark.vcr()
def test_be_client_error(
    db,
    admin_client,
    snapshot,
    vcr_config,
    egrids,
    be_data_sources,
):
    response = admin_client.get(
        reverse("gis-data"),
        data={
            "egrids": egrids,
        },
    )

    assert response.status_code == status.HTTP_400_BAD_REQUEST
    snapshot.assert_match(response.json())
