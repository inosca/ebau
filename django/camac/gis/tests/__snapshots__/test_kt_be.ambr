# serializer version: 1
# name: test_be_client[CH643546955207,CH843546955632]
  dict({
    'data': dict({
      'belasteter-standort': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'belasteter-standort',
        'value': 'belasteter-standort-ja',
      }),
      'gebiet-mit-archaeologischen-objekten': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'gebiet-mit-archaeologischen-objekten',
        'value': 'gebiet-mit-archaeologischen-objekten-nein',
      }),
      'gebiet-mit-naturgefahren': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'gebiet-mit-naturgefahren',
        'value': 'gebiet-mit-naturgefahren-nein',
      }),
      'gewaesserschutzbereich-v2': dict({
        'hidden': False,
        'label': 'gewaesserschutzbereich-v2',
        'value': list([
          dict({
            'displayValue': 'ueb',
            'value': 'gewaesserschutzbereich-v2-ueb',
          }),
          dict({
            'displayValue': 'ao',
            'value': 'gewaesserschutzbereich-v2-ao',
          }),
        ]),
      }),
      'handelt-es-sich-um-ein-baudenkmal': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'handelt-es-sich-um-ein-baudenkmal',
        'value': 'handelt-es-sich-um-ein-baudenkmal-ja',
      }),
      'naturschutz': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'naturschutz',
        'value': 'naturschutz-ja',
      }),
      'objekt-des-besonderen-landschaftsschutzes': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'objekt-des-besonderen-landschaftsschutzes',
        'value': 'objekt-des-besonderen-landschaftsschutzes-nein',
      }),
    }),
  })
# ---
# name: test_be_client[CH643546955207,CH851446093521,CH673533354667]
  dict({
    'data': dict({
      'belasteter-standort': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'belasteter-standort',
        'value': 'belasteter-standort-ja',
      }),
      'gebiet-mit-archaeologischen-objekten': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-archaeologischen-objekten',
        'value': 'gebiet-mit-archaeologischen-objekten-ja',
      }),
      'gebiet-mit-naturgefahren': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-naturgefahren',
        'value': 'gebiet-mit-naturgefahren-ja',
      }),
      'gewaesserschutzbereich-v2': dict({
        'hidden': False,
        'label': 'gewaesserschutzbereich-v2',
        'value': list([
          dict({
            'displayValue': 'ueb',
            'value': 'gewaesserschutzbereich-v2-ueb',
          }),
          dict({
            'displayValue': 'ao',
            'value': 'gewaesserschutzbereich-v2-ao',
          }),
          dict({
            'displayValue': 'au',
            'value': 'gewaesserschutzbereich-v2-au',
          }),
        ]),
      }),
      'handelt-es-sich-um-ein-baudenkmal': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'handelt-es-sich-um-ein-baudenkmal',
        'value': 'handelt-es-sich-um-ein-baudenkmal-ja',
      }),
      'naturschutz': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'naturschutz',
        'value': 'naturschutz-ja',
      }),
      'nutzungszone': dict({
        'hidden': False,
        'label': 'nutzungszone',
        'value': 'Gewerbezone_G2, Grünzone_GrZ, Wohnzone 1 E1, Wohnzone 2 E2, Zone für öffentliche Nutzung ZöN 7, Zone_für_öffentliche_Nutzung, Zone_für_öffentliche_Nutzung_B, Zone_für_öffentliche_Nutzung_C, Zone_für_öffentliche_Nutzung_D',
      }),
      'objekt-des-besonderen-landschaftsschutzes': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'objekt-des-besonderen-landschaftsschutzes',
        'value': 'objekt-des-besonderen-landschaftsschutzes-nein',
      }),
      'ueberbauungsordnung': dict({
        'hidden': False,
        'label': 'ueberbauungsordnung',
        'value': 'ZPP_2_"Aarolina", ZPP_5_"Chräjeninsel", ÜO_"Arolina", ÜO_"Chräjeninsel"',
      }),
    }),
  })
# ---
# name: test_be_client[CH643546955207,CH851446093521]
  dict({
    'data': dict({
      'belasteter-standort': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'belasteter-standort',
        'value': 'belasteter-standort-ja',
      }),
      'gebiet-mit-archaeologischen-objekten': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-archaeologischen-objekten',
        'value': 'gebiet-mit-archaeologischen-objekten-ja',
      }),
      'gebiet-mit-naturgefahren': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'gebiet-mit-naturgefahren',
        'value': 'gebiet-mit-naturgefahren-nein',
      }),
      'gewaesserschutzbereich-v2': dict({
        'hidden': False,
        'label': 'gewaesserschutzbereich-v2',
        'value': list([
          dict({
            'displayValue': 'ueb',
            'value': 'gewaesserschutzbereich-v2-ueb',
          }),
          dict({
            'displayValue': 'ao',
            'value': 'gewaesserschutzbereich-v2-ao',
          }),
          dict({
            'displayValue': 'au',
            'value': 'gewaesserschutzbereich-v2-au',
          }),
        ]),
      }),
      'handelt-es-sich-um-ein-baudenkmal': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'handelt-es-sich-um-ein-baudenkmal',
        'value': 'handelt-es-sich-um-ein-baudenkmal-ja',
      }),
      'naturschutz': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'naturschutz',
        'value': 'naturschutz-ja',
      }),
      'nutzungszone': dict({
        'hidden': False,
        'label': 'nutzungszone',
        'value': 'Wohnzone 1 E1, Wohnzone 2 E2, Zone für öffentliche Nutzung ZöN 7',
      }),
      'objekt-des-besonderen-landschaftsschutzes': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'objekt-des-besonderen-landschaftsschutzes',
        'value': 'objekt-des-besonderen-landschaftsschutzes-nein',
      }),
    }),
  })
# ---
# name: test_be_client[CH643546955207]
  dict({
    'data': dict({
      'belasteter-standort': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'belasteter-standort',
        'value': 'belasteter-standort-ja',
      }),
      'gebiet-mit-archaeologischen-objekten': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'gebiet-mit-archaeologischen-objekten',
        'value': 'gebiet-mit-archaeologischen-objekten-nein',
      }),
      'gebiet-mit-naturgefahren': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'gebiet-mit-naturgefahren',
        'value': 'gebiet-mit-naturgefahren-nein',
      }),
      'gewaesserschutzbereich-v2': dict({
        'hidden': False,
        'label': 'gewaesserschutzbereich-v2',
        'value': list([
          dict({
            'displayValue': 'ueb',
            'value': 'gewaesserschutzbereich-v2-ueb',
          }),
          dict({
            'displayValue': 'ao',
            'value': 'gewaesserschutzbereich-v2-ao',
          }),
        ]),
      }),
      'handelt-es-sich-um-ein-baudenkmal': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'handelt-es-sich-um-ein-baudenkmal',
        'value': 'handelt-es-sich-um-ein-baudenkmal-ja',
      }),
      'naturschutz': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'naturschutz',
        'value': 'naturschutz-ja',
      }),
      'objekt-des-besonderen-landschaftsschutzes': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'objekt-des-besonderen-landschaftsschutzes',
        'value': 'objekt-des-besonderen-landschaftsschutzes-nein',
      }),
    }),
  })
# ---
# name: test_be_client[CH673533354667,CH643546955207]
  dict({
    'data': dict({
      'belasteter-standort': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'belasteter-standort',
        'value': 'belasteter-standort-ja',
      }),
      'gebiet-mit-archaeologischen-objekten': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-archaeologischen-objekten',
        'value': 'gebiet-mit-archaeologischen-objekten-ja',
      }),
      'gebiet-mit-naturgefahren': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-naturgefahren',
        'value': 'gebiet-mit-naturgefahren-ja',
      }),
      'gewaesserschutzbereich-v2': dict({
        'hidden': False,
        'label': 'gewaesserschutzbereich-v2',
        'value': list([
          dict({
            'displayValue': 'ueb',
            'value': 'gewaesserschutzbereich-v2-ueb',
          }),
          dict({
            'displayValue': 'ao',
            'value': 'gewaesserschutzbereich-v2-ao',
          }),
          dict({
            'displayValue': 'au',
            'value': 'gewaesserschutzbereich-v2-au',
          }),
        ]),
      }),
      'handelt-es-sich-um-ein-baudenkmal': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'handelt-es-sich-um-ein-baudenkmal',
        'value': 'handelt-es-sich-um-ein-baudenkmal-ja',
      }),
      'naturschutz': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'naturschutz',
        'value': 'naturschutz-ja',
      }),
      'nutzungszone': dict({
        'hidden': False,
        'label': 'nutzungszone',
        'value': 'Gewerbezone_G2, Grünzone_GrZ, Zone_für_öffentliche_Nutzung, Zone_für_öffentliche_Nutzung_B, Zone_für_öffentliche_Nutzung_C, Zone_für_öffentliche_Nutzung_D',
      }),
      'objekt-des-besonderen-landschaftsschutzes': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'objekt-des-besonderen-landschaftsschutzes',
        'value': 'objekt-des-besonderen-landschaftsschutzes-nein',
      }),
      'ueberbauungsordnung': dict({
        'hidden': False,
        'label': 'ueberbauungsordnung',
        'value': 'ZPP_2_"Aarolina", ZPP_5_"Chräjeninsel", ÜO_"Arolina", ÜO_"Chräjeninsel"',
      }),
    }),
  })
# ---
# name: test_be_client[CH673533354667,CH851446093521]
  dict({
    'data': dict({
      'belasteter-standort': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'belasteter-standort',
        'value': 'belasteter-standort-ja',
      }),
      'gebiet-mit-archaeologischen-objekten': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-archaeologischen-objekten',
        'value': 'gebiet-mit-archaeologischen-objekten-ja',
      }),
      'gebiet-mit-naturgefahren': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-naturgefahren',
        'value': 'gebiet-mit-naturgefahren-ja',
      }),
      'gewaesserschutzbereich-v2': dict({
        'hidden': False,
        'label': 'gewaesserschutzbereich-v2',
        'value': list([
          dict({
            'displayValue': 'ao',
            'value': 'gewaesserschutzbereich-v2-ao',
          }),
          dict({
            'displayValue': 'au',
            'value': 'gewaesserschutzbereich-v2-au',
          }),
        ]),
      }),
      'handelt-es-sich-um-ein-baudenkmal': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'handelt-es-sich-um-ein-baudenkmal',
        'value': 'handelt-es-sich-um-ein-baudenkmal-ja',
      }),
      'naturschutz': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'naturschutz',
        'value': 'naturschutz-nein',
      }),
      'nutzungszone': dict({
        'hidden': False,
        'label': 'nutzungszone',
        'value': 'Gewerbezone_G2, Grünzone_GrZ, Wohnzone 1 E1, Wohnzone 2 E2, Zone für öffentliche Nutzung ZöN 7, Zone_für_öffentliche_Nutzung, Zone_für_öffentliche_Nutzung_B, Zone_für_öffentliche_Nutzung_C, Zone_für_öffentliche_Nutzung_D',
      }),
      'objekt-des-besonderen-landschaftsschutzes': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'objekt-des-besonderen-landschaftsschutzes',
        'value': 'objekt-des-besonderen-landschaftsschutzes-nein',
      }),
      'ueberbauungsordnung': dict({
        'hidden': False,
        'label': 'ueberbauungsordnung',
        'value': 'ZPP_2_"Aarolina", ZPP_5_"Chräjeninsel", ÜO_"Arolina", ÜO_"Chräjeninsel"',
      }),
    }),
  })
# ---
# name: test_be_client[CH673533354667]
  dict({
    'data': dict({
      'belasteter-standort': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'belasteter-standort',
        'value': 'belasteter-standort-ja',
      }),
      'gebiet-mit-archaeologischen-objekten': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-archaeologischen-objekten',
        'value': 'gebiet-mit-archaeologischen-objekten-ja',
      }),
      'gebiet-mit-naturgefahren': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-naturgefahren',
        'value': 'gebiet-mit-naturgefahren-ja',
      }),
      'gewaesserschutzbereich-v2': dict({
        'hidden': False,
        'label': 'gewaesserschutzbereich-v2',
        'value': list([
          dict({
            'displayValue': 'au',
            'value': 'gewaesserschutzbereich-v2-au',
          }),
        ]),
      }),
      'handelt-es-sich-um-ein-baudenkmal': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'handelt-es-sich-um-ein-baudenkmal',
        'value': 'handelt-es-sich-um-ein-baudenkmal-nein',
      }),
      'naturschutz': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'naturschutz',
        'value': 'naturschutz-nein',
      }),
      'nutzungszone': dict({
        'hidden': False,
        'label': 'nutzungszone',
        'value': 'Gewerbezone_G2, Grünzone_GrZ, Zone_für_öffentliche_Nutzung, Zone_für_öffentliche_Nutzung_B, Zone_für_öffentliche_Nutzung_C, Zone_für_öffentliche_Nutzung_D',
      }),
      'objekt-des-besonderen-landschaftsschutzes': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'objekt-des-besonderen-landschaftsschutzes',
        'value': 'objekt-des-besonderen-landschaftsschutzes-nein',
      }),
      'ueberbauungsordnung': dict({
        'hidden': False,
        'label': 'ueberbauungsordnung',
        'value': 'ZPP_2_"Aarolina", ZPP_5_"Chräjeninsel", ÜO_"Arolina", ÜO_"Chräjeninsel"',
      }),
    }),
  })
# ---
# name: test_be_client[CH851446093521]
  dict({
    'data': dict({
      'belasteter-standort': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'belasteter-standort',
        'value': 'belasteter-standort-nein',
      }),
      'gebiet-mit-archaeologischen-objekten': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'gebiet-mit-archaeologischen-objekten',
        'value': 'gebiet-mit-archaeologischen-objekten-ja',
      }),
      'gebiet-mit-naturgefahren': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'gebiet-mit-naturgefahren',
        'value': 'gebiet-mit-naturgefahren-nein',
      }),
      'gewaesserschutzbereich-v2': dict({
        'hidden': False,
        'label': 'gewaesserschutzbereich-v2',
        'value': list([
          dict({
            'displayValue': 'ao',
            'value': 'gewaesserschutzbereich-v2-ao',
          }),
          dict({
            'displayValue': 'au',
            'value': 'gewaesserschutzbereich-v2-au',
          }),
        ]),
      }),
      'handelt-es-sich-um-ein-baudenkmal': dict({
        'displayValue': 'ja',
        'hidden': False,
        'label': 'handelt-es-sich-um-ein-baudenkmal',
        'value': 'handelt-es-sich-um-ein-baudenkmal-ja',
      }),
      'naturschutz': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'naturschutz',
        'value': 'naturschutz-nein',
      }),
      'nutzungszone': dict({
        'hidden': False,
        'label': 'nutzungszone',
        'value': 'Wohnzone 1 E1, Wohnzone 2 E2, Zone für öffentliche Nutzung ZöN 7',
      }),
      'objekt-des-besonderen-landschaftsschutzes': dict({
        'displayValue': 'nein',
        'hidden': False,
        'label': 'objekt-des-besonderen-landschaftsschutzes',
        'value': 'objekt-des-besonderen-landschaftsschutzes-nein',
      }),
    }),
  })
# ---
# name: test_be_client_error[doesntexist]
  list([
    'No polygon found',
  ])
# ---
# name: test_be_client_error[emptygis]
  list([
    'No polygon found',
  ])
# ---
# name: test_be_client_error[emptypolygon]
  list([
    'No polygon found',
  ])
# ---
